rpc+tapcli: Implement method to decode proofs into human-readable format

habibitcoin opened the pull request:
Related to https://github.com/lightninglabs/taro/issues/180

Would be nice to be able to view proofs in human readable format (https://github.com/lightninglabs/taro/blob/main/proof/proof_test.go#L127); this will make it easier for individuals to wrap their mind around the mental model of proofs, and what they contain, and ideally accelerate universe development

Able to access some of the raw data!

edit: here is sample output as of now:
```
taro % tarocli proofs decode --proof_file habibtaro_usd_jan2.json
{
    "proof_index": "0",
    "asset": {
        "version": 0,
        "asset_genesis": {
            "genesis_point": "3dea4202dc504302a68fa598ed6ade31f83fdc3a58b095f4d6259587fa48dea0:1",
            "name": "habibtaro_usd_jan_noem23",
            "meta": "68616269627461726f2e636f6d5f6a616e5f6e6f656d3233",
            "asset_id": "8074f97dc3f2a5c3375ddaccf66ffc0bd1fbdd0b8261ee536a1bc9653c0d9f0b",
            "output_index": 0,
            "genesis_bootstrap_info": "a0de48fa879525d6f495b0583adc3ff831de6aed98a58fa6024350dc0242ea3d000000011868616269627461726f5f7573645f6a616e5f6e6f656d32331868616269627461726f2e636f6d5f6a616e5f6e6f656d32330000000000",
            "version": 0
        },
        "asset_type": "NORMAL",
        "amount": "21000000",
        "lock_time": 0,
        "relative_lock_time": 0,
        "script_version": 0,
        "script_key": "020c7b5c4abd3b7cbe48a1cb8193312233924f22cfdc07e9c64a7f867cca3b5a3e",
        "asset_group": {
            "raw_group_key": null,
            "tweaked_group_key": "0253c2de331021b68632d5cb76110d4a65e4436f7762e47b8af058641012694110",
            "asset_id_sig": "cb61c6c21047b288e116880cb6abf941c754127a113e081e14b65cf834f445325817c4be9cdd44703e05b3a0d60adf10392b454814f89635e9339ae0f55fced8"
        },
        "chain_anchor": {
            "anchor_tx": "02000000000101a0de48fa879525d6f495b0583adc3ff831de6aed98a58fa6024350dc0242ea3d0100000000ffffffff02e803000000000000225120593814466c8ae931db5a335f62737d1f5979f7f1fa0d3b75de38c1bc356b584927d2570000000000225120951d666ae736743e5dfe8eaeaad80039f689df429bf0220dd245d1938dc21cfb0140a392cedef28b6efdb26dc31f354a8633ca6301aef94e8f55ed772f6a2fbb9673735a0f80a68d10dc57980560f94a98d1fc2b480b447818a6039ac0229139eea500000000",
            "anchor_txid": "986c8f9263aa5a0d1d6733019bc8660216b3e6a3380c38b73e39811881c677bd",
            "anchor_block_hash": "000000000004479a4e6a788f3994069b2c15b3b61b00673fd8a099323f4f6e78",
            "anchor_outpoint": "986c8f9263aa5a0d1d6733019bc8660216b3e6a3380c38b73e39811881c677bd:0",
            "internal_key": null
        }
    },
    "tx_merkle_proof": "06912174fe872bb334e87038d02313e51f85e76442f79b73b04e0c10b3330c84c2b2dc73a43a0cadf17126b05122e6a54f2375572e5bb4f4d02b9fd7b2d3741ede711f85425cd99a2fd0a2b7b735de769274f44d3564105fc2829821c69f37e111be624fc946a60f91f90b4d7462581bca9790ad6099e1effb271733a9ec7231fbd04772186e38be31da8c88d457753455778ee96ab60a76f44c1e4065e374a3e06bfc6be5cc665baf7d81dd617ec34b0e6cd91252f8a46221763aedb35872fef738",
    "inclusion_proof": "00040000000001210285127fca8f36e75b0508e6a6a561ab5db97003523f227596bc4f7be436a9b7a402c400490001000120343934bd45c74d19ae43f6034b639f70232264ad03d7b02555b5a3f478cfbe9202220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0177000100017200026b9241c6ff865f4be097261e3a4170e7ceceb7e51e7565b9a5b692fc6a9ad3d80000000001406f40e62fa07552a32c0fc1179e234f024ed3d833b924c773810e649d3da1ebae8b8d000000000280de80ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff5f",
    "exclusion_proofs": [
        "000400000001012102c17d96c48980a0f48932ac66819e6ddc86cce68edaae0179d66d94326ce130160303020101"
    ]
}
```

Roasbeef commented:
Answering the questions you posted:

>  Should we only return the last proof?

I think so. It's also possible to pass in an index, then we can seek to the desired index using this method: https://github.com/lightninglabs/taro/blob/main/proof/file.go#L203-L205

> Should we return any of the merkle trees/additional information? Block header? Additional inputs?

I think the block header/height information would be useful. Then maybe a verbose option to include the serialized Bitcoin merkle proof as well as the series of _inclusion_ (and exclusion) proofs? 

> Is the anchor tx available in proofs?
Yes, it's this field: https://github.com/lightninglabs/taro/blob/77b69191fc60c57135c70e25b338d5d401bfbb2b/proof/proof.go#L40-L42

> Am I supposed to commit the go.mod/go.sum files?

Yeah, this commits the set of dependancies.
 
> Think I need to fix taro.swagger.json, maybe missing other generated files?
> Am I supposed to update the protoc-gen-go and protoc versions or leave that out?

If you do `make rpc`, then it should handle creating all the auto generated file diffs. 

I'd leave those versions the same, since we pin to the same version so everyone generates the same set of protos.

Roasbeef commented on go.mod:
```diff
@@ -27,10 +27,10 @@ require (
-	golang.org/x/term v0.0.0-20201126162022-7de9c90e9dd1
-	google.golang.org/grpc v1.39.0
-	google.golang.org/protobuf v1.27.1
+	golang.org/x/sync v0.0.0-20220722155255-886fb9371eb4
```
Which version of Go are you running? Did you do a `go mod tidy` that triggered these changes?

habibitcoin replied:
Figured that out, was on go1.19

Went to go1.18, and then these files were unmodified

Roasbeef commented on cmd/tarocli/proofs.go:
```diff
@@ -20,6 +20,7 @@ var proofCommands = []cli.Command{
 		Category:  "Proofs",
 		Subcommands: []cli.Command{
 			verifyProofCommand,
+			decodeProofCommand,
```
So I think we can probably split up the diff into a few commits:
  * Add the new proto fields + run the code generation 
  * Implement the decode RPC
  * Add the new CLI command

dstadulis commented:
@habibitcoin Any update on when we can expect this?

habibitcoin commented:
Thanks for the review @Roasbeef ! All makes sense

@dstadulis expecting to complete this over Thanksgiving holiday at latest

lightninglabs-deploy commented:
@jharveyb: review reminder
@habibitcoin, remember to re-request review from reviewers when ready

habibitcoin commented:
@Roasbeef I'm having a bit of trouble on 
>Then maybe a verbose option to include the serialized Bitcoin merkle proof as well as the series of inclusion (and exclusion) proofs?

1. What is the serialized Bitcoin merkle proof?

2. Would these fields ideally be output as bytes? If so, is this the proper way to do so:
```
var inclusionProofBuf bytes.Buffer
	if err := decodedProof.InclusionProof.Encode(&inclusionProofBuf); err != nil {
		return nil, fmt.Errorf("unable to encode inclusion proof: %w",
			err)
	}
```

3. For the exclusion proofs, would that be an array of bytes? Would that be done by:
```
repeated bytes exclusion_proofs
```

I believe everything else is implemented! Example output below:

![Screen Shot 2022-11-27 at 12 56 21 PM](https://user-images.githubusercontent.com/114780316/204151879-7693e3be-4200-4f29-be9a-207c943633fc.png)

guggero commented:
> 1. What is the serialized Bitcoin merkle proof?

The field `TxMerkleProof` of type `TxMerkleProof`.

> 2. Would these fields ideally be output as bytes? If so, is this the proper way to do so:

Yes, that looks good. Seems like all the fields mentioned (`TxMerkleProof`,  `InclusionProof` and `ExclusionProofs`) have an `Encode` method. 

> For the exclusion proofs, would that be an array of bytes? Would that be done by:

Yes, `repeated bytes exclusion_proofs` sounds good on the RPC level.

> Have a bug with block_height output*

Yes, the `block_height` should be a simple block number. But that information is not included in the proof, so you can't print it.

lightninglabs-deploy commented:
@roasbeef: review reminder

habibitcoin commented:
Latest output:

```
{
    "proof_index": "0",
    "anchor_outpoint": "e3dcdfe1f03821430a13742189af3a8c451635b113e852f690b4b59dd864a643:1",
    "anchor_tx": "063280c2cd2a1c1cf42cf2fc3fb21cd6660a097e22dadc8db5f733ce36216a3d",
    "block_header": "000000000000000d14419cb8be4dabdea503ed194be79e2c6063d2aefb1fb646",
    "block_height": "0000000000000015879d4297902849c42bd5f8b17cc48c43b9150321f7f845b2",
    "asset": {
        "version": 0,
        "asset_genesis": {
            "genesis_point": "e3dcdfe1f03821430a13742189af3a8c451635b113e852f690b4b59dd864a643:1",
            "name": "habibtaro",
            "meta": "68616269627461726f2e636f6d",
            "asset_id": "f30905681df0b8b6122ba38fbf3711d01ede2ba678cfbd4454486d7027972501",
            "output_index": 0,
            "genesis_bootstrap_info": "43a664d89db5b490f652e813b13516458c3aaf892174130a432138f0e1dfdce3000000010968616269627461726f0d68616269627461726f2e636f6d0000000000",
            "version": 0
        },
        "asset_type": "NORMAL",
        "amount": "21000000",
        "lock_time": 0,
        "relative_lock_time": 0,
        "script_version": 0,
        "script_key": "0206b93c43b8a85385e1f6021108901ef99eeb5a69378ec805dea8abc4a3fd868c",
        "asset_group": {
            "raw_group_key": null,
            "tweaked_group_key": "0337e043736df85040e97def7c8cdb55dc97f553348d24058167622b585630e178",
            "asset_id_sig": "a0e4d260effd6a4babaf8a00d6534e81a2801f6723d99af768cb1ca17202bf77525f16d04e70691bc8941fedf8bf06ba9f01ce95f19a19178ceaad39f8a29ba3"
        },
        "chain_anchor": {
            "anchor_tx": "0200000000010143a664d89db5b490f652e813b13516458c3aaf892174130a432138f0e1dfdce30100000000ffffffff02e803000000000000225120e1e23ab4aa5f26d54dd24be4401e8eb0e560a09a94a428481408fd53e404ae0bcdc1580000000000160014c7cfecd42c55786054ec4eaad71b2f62e68428fd024730440220321d45c17beb16a68f7ba8a23cf386de149f3b4b52082b58d555f2e08a9efbde02202a3ef23f5b1e8cd6241ae0ec3c922e48f858281cf7a9462d7e0bc4c218cff0350121030c999ac2ee7695f855f525b1a24e97d53ada26dfa4df83cbf40df173e38664d200000000",
            "anchor_txid": "3d6a2136ce33f7b58ddcda227e090a66d61cb23ffcf22cf41c1c2acdc2803206",
            "anchor_block_hash": "46b61ffbaed263602c9ee74b19ed03a5deab4dbeb89c41140d00000000000000",
            "anchor_outpoint": "e3dcdfe1f03821430a13742189af3a8c451635b113e852f690b4b59dd864a643:1",
            "internal_key": null
        }
    },
    "tx_merkle_proof": "038082936f2a9190fb7de3ef010e9a3354ca6e08aa2d88d1286c3a437b126084cfa3ab6bdd46c6fd76ebee5a65b106daf9acc26646299f4cf3468a49af16308d93d73b19aafc73920016b43b6458b9f63dfdd6329929052cf313817e9aeaf76f0d03",
    "inclusion_proof": "0004000000000121021b816b11a35a95aae3c59b314895ddd22e0f68ce6b1764b05d9f3928185672ad02740049000100012046690b2765ce8d60837116cd0150f6defd733b66df9f6d6caa47883fcd09bc7202220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff012700010001220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
    "exclusion_proofs": [
    ]
}
```

habibitcoin commented:
Also thinking to just remove block height because there doesnt appear to be any easy way to extract this information

Roasbeef commented:
> Also thinking to just remove block height because there doesnt appear to be any easy way to extract this information

Should just be available as a field: https://github.com/lightninglabs/taro/blob/f6a581b6395fe779db7cbb5a4536a39bb0440dbe/proof/file.go#L317-L318

lightninglabs-deploy commented:
@roasbeef: review reminder

lightninglabs-deploy commented:
@roasbeef: review reminder

lightninglabs-deploy commented:
@roasbeef: review reminder

Roasbeef commented:
Is this ready for another round @habibitcoin ?

lightninglabs-deploy commented:
@habibitcoin, remember to re-request review from reviewers when ready

habibitcoin commented:
Same issue https://github.com/lightninglabs/taro/pull/230 for AnchorBlockHash. Included that.

Ready for final review @Roasbeef !

```
taro % tarocli proofs decode --proof_file habibtaro_usd_jan2.json
{
    "proof_index": "0",
    "asset": {
        "version": 0,
        "asset_genesis": {
            "genesis_point": "3dea4202dc504302a68fa598ed6ade31f83fdc3a58b095f4d6259587fa48dea0:1",
            "name": "habibtaro_usd_jan_noem23",
            "meta": "68616269627461726f2e636f6d5f6a616e5f6e6f656d3233",
            "asset_id": "8074f97dc3f2a5c3375ddaccf66ffc0bd1fbdd0b8261ee536a1bc9653c0d9f0b",
            "output_index": 0,
            "genesis_bootstrap_info": "a0de48fa879525d6f495b0583adc3ff831de6aed98a58fa6024350dc0242ea3d000000011868616269627461726f5f7573645f6a616e5f6e6f656d32331868616269627461726f2e636f6d5f6a616e5f6e6f656d32330000000000",
            "version": 0
        },
        "asset_type": "NORMAL",
        "amount": "21000000",
        "lock_time": 0,
        "relative_lock_time": 0,
        "script_version": 0,
        "script_key": "020c7b5c4abd3b7cbe48a1cb8193312233924f22cfdc07e9c64a7f867cca3b5a3e",
        "asset_group": {
            "raw_group_key": null,
            "tweaked_group_key": "0253c2de331021b68632d5cb76110d4a65e4436f7762e47b8af058641012694110",
            "asset_id_sig": "cb61c6c21047b288e116880cb6abf941c754127a113e081e14b65cf834f445325817c4be9cdd44703e05b3a0d60adf10392b454814f89635e9339ae0f55fced8"
        },
        "chain_anchor": {
            "anchor_tx": "02000000000101a0de48fa879525d6f495b0583adc3ff831de6aed98a58fa6024350dc0242ea3d0100000000ffffffff02e803000000000000225120593814466c8ae931db5a335f62737d1f5979f7f1fa0d3b75de38c1bc356b584927d2570000000000225120951d666ae736743e5dfe8eaeaad80039f689df429bf0220dd245d1938dc21cfb0140a392cedef28b6efdb26dc31f354a8633ca6301aef94e8f55ed772f6a2fbb9673735a0f80a68d10dc57980560f94a98d1fc2b480b447818a6039ac0229139eea500000000",
            "anchor_txid": "986c8f9263aa5a0d1d6733019bc8660216b3e6a3380c38b73e39811881c677bd",
            "anchor_block_hash": "000000000004479a4e6a788f3994069b2c15b3b61b00673fd8a099323f4f6e78",
            "anchor_outpoint": "986c8f9263aa5a0d1d6733019bc8660216b3e6a3380c38b73e39811881c677bd:0",
            "internal_key": null
        }
    },
    "tx_merkle_proof": "06912174fe872bb334e87038d02313e51f85e76442f79b73b04e0c10b3330c84c2b2dc73a43a0cadf17126b05122e6a54f2375572e5bb4f4d02b9fd7b2d3741ede711f85425cd99a2fd0a2b7b735de769274f44d3564105fc2829821c69f37e111be624fc946a60f91f90b4d7462581bca9790ad6099e1effb271733a9ec7231fbd04772186e38be31da8c88d457753455778ee96ab60a76f44c1e4065e374a3e06bfc6be5cc665baf7d81dd617ec34b0e6cd91252f8a46221763aedb35872fef738",
    "inclusion_proof": "00040000000001210285127fca8f36e75b0508e6a6a561ab5db97003523f227596bc4f7be436a9b7a402c400490001000120343934bd45c74d19ae43f6034b639f70232264ad03d7b02555b5a3f478cfbe9202220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0177000100017200026b9241c6ff865f4be097261e3a4170e7ceceb7e51e7565b9a5b692fc6a9ad3d80000000001406f40e62fa07552a32c0fc1179e234f024ed3d833b924c773810e649d3da1ebae8b8d000000000280de80ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff5f",
    "exclusion_proofs": [
        "000400000001012102c17d96c48980a0f48932ac66819e6ddc86cce68edaae0179d66d94326ce130160303020101"
    ]
}
```

habibitcoin commented:
Also 1 last question, wondering if I should include this is the same or a new PR: meta should not be displayed as bytes. It is sent in via CLI as a string, so should be output as the string:

```
"meta": "habibtaro.com_jan_noem23",
versus
 "meta": "68616269627461726f2e636f6d5f6a616e5f6e6f656d3233",
```

String seems much more useful?

lightninglabs-deploy commented:
@roasbeef: review reminder

Roasbeef commented:
> Also 1 last question, wondering if I should include this is the same or a new PR: meta should not be displayed as bytes. It is sent in via CLI as a string, so should be output as the string:


I think for now can have it assume things are displayed as bytes, I'm making some changes to how the meta bytes are presented/transmitted in this PR: #249.

Roasbeef commented on rpcserver.go:
```diff
@@ -392,7 +392,7 @@ func (r *rpcServer) fetchRpcAssets(ctx context.Context) (
 				AnchorTx:        anchorTxBytes,
 				AnchorTxid:      a.AnchorTxid.String(),
-				AnchorBlockHash: a.AnchorBlockHash[:],
+				AnchorBlockHash: a.AnchorBlockHash.String(),
```
👍

Roasbeef commented on tarorpc/taro.proto:
```diff
@@ -508,8 +514,24 @@ message ProofFile {
+
+message DecodeProofRequest {
+    bytes raw_proof = 1;
+    int64 proof_index = 2;
```
Which lets you seek to a certain proof within the greater multi-proof? Does a value of -1 or 0 here mean "verify the entire thing"? Or is the idea it only does things in fragments?

habibitcoin replied:
Ahh so DecodeProof will always only decode 1 proof, and same with VerifyProof

But VerifyProof will continue to verify the entire proof. Will make comments clear

guggero replied:
The "defaults to the latest proof" is incorrect here. Because "default" means you don't specify the value, in which case gRPC will use `0`. So by default it would verify the genesis proof (first in the file).

habibitcoin replied:
@guggero doesn't this https://github.com/lightninglabs/taro/pull/181/files#diff-9bf3925007bcf9962c1a4b80fe74052ee85dda0a9b25915ddcb78b8f0ac075eaR90 create the default value of -1?

When testing, I observed it working properly to default to latest proof instead of genesis

guggero replied:
Yes, for the CLI. But people will also only interact with Tarod on the RPC (and will only look at the generated API docs), so the RPC description is important too.

habibitcoin replied:
Ohhh yeah good point.. hmmmmm...

Will give this some thought, and happy to take suggestions.

Maybe 1 direction to go is having the index start at 1 instead of 0? Wdyt?

guggero replied:
No, I think that's just ugly. I think @Roasbeef suggested this already (maybe in a different context), we could use the `oneof` pattern that gRPC offers here:
```
oneof filter {
  bool all_proofs = 2;
  uint32 index = 3;
}
```

That way you can specify either `all_proofs` or set an index.

habibitcoin replied:
If all_proofs was set to true, would you expect to see _all_ the proofs decoded in the response? 

Must say I'm not a fan of the oneof approach either, but I can't think of anything better yet either

Because if the user doesn't send all_proofs, it would default to false right?

habibitcoin replied:
I think it can work if the bool is partialverify, wdyt? And we have the check for index > 0 as the first case in the switch.

So if a user doesn't set either, we assume we want a fullverify.

If the user only provides us an index > 0, we respect that and ignore partialverify

If a user provides us an index of 0, we will then look at partialverify to determine if they want the latest proof, or if partialverify is true, we know they want the 0 proof

habibitcoin replied:
Or actually I think I need to understand oneof better https://www.tutorialspoint.com/protobuf/protobuf_compound_data_types.htm

Do I understand correctly that if _both_ fields are provided in a group by, the 2nd field will take precedence because it overwrites the first?

habibitcoin replied:
Perhaps just making index optional is a route to go here? https://stackoverflow.com/questions/42622015/how-to-define-an-optional-field-in-protobuf-3

habibitcoin replied:
Ok sorry for the confusion, I think I get it now.
```
oneof filter {
  bool all_proofs = 2;
  uint32 index = 3;
}
```

So we expect index to be the most common field that people would _explicitly_ send. If they do not explicitly send an index, we can default to all_proofs

```
switch filter := in.Filter.(type) {
        case *Index:
		We use the provided index and perform partial_verify

	default:
		We do a full verification
	}
```

habibitcoin replied:
This is what I have

Lmk your thoughts @guggero 

```
	var proofIdx int64
	switch filter := in.Filter.(type) {
	case *tarorpc.DecodeProofRequest_ProofIndex:
		proofIdx = filter.ProofIndex
	default:
		proofIdx = -1
	}

	decodedProof, err := marshalProofFile(proofFile, proofIdx)
	if err != nil {
		return nil, fmt.Errorf("unable to marshal proof: %w", err)
	}
	```

habibitcoin replied:
And then inside tarocli/proofs.go:

```
req := &tarorpc.DecodeProofRequest{
		RawProof: rawFile,
	}

	if !ctx.Bool(allProofsName) {
		req.Filter = &tarorpc.DecodeProofRequest_ProofIndex{
			ProofIndex: ctx.Int64(proofIndexName),
		}
	} else {
		req.Filter = &tarorpc.DecodeProofRequest_AllProofs{
			AllProofs: true,
		}
	}

	resp, err := client.DecodeProof(ctxc, req)
	if err != nil {
		return fmt.Errorf("unable to verify file: %w", err)
	}
	```
	
	Let me know if this works, as I've tested the solution and it works properly over REST!

guggero replied:
Yes, this looks good. Though I think I wanted to call it `latest_proof` instead of `all_proofs`, was answering in a rush without reloading all context, sorry.

Roasbeef commented on tarorpc/taro.proto:
```diff
@@ -508,8 +514,24 @@ message ProofFile {
+    int64 proof_index = 2;
+}
+
+message DecodeProofResponse {
```
Can potentially expose more data here, but I think this is good as an initial set.

positiveblue replied:
nit: split this in two different messages

```
message DecodedProof {
// decoded proof fields
}

message DecodeProofResponse {
    DecodedProof decoded_proof = 1
}

// VerifyProofResponse does not use anything called "response" then, that is supposed to be for method
// responses
message VerifyProofResponse {
    bool valid = 1;
    DecodedProof decoded_proof = 2;
}
```

positiveblue replied:
After reviewing the implementation I think we should also add `number_of_proofs` to the `DecodeProofResponse`. By default I want to get info about the last one, but I would also want to know how many are there in case I want to use the index parameter.

Roasbeef commented on rpcserver.go:
```diff
@@ -791,7 +796,7 @@ func (r *rpcServer) DecodeAddr(_ context.Context,
 // at the specified genesis point.
 func (r *rpcServer) VerifyProof(ctx context.Context,
-	in *tarorpc.ProofFile) (*tarorpc.ProofVerifyResponse, error) {
+	in *tarorpc.ProofFile) (*tarorpc.VerifyProofResponse, error) {
```
+1 for the naming consistency here.

Roasbeef commented on rpcserver.go:
```diff
@@ -809,15 +814,151 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+}
+
+// marshalProofFile turns a proof file into an RPC DecodeProofResponse.
+func marshalProofFile(proofFile proof.File, index int64) (*tarorpc.DecodeProofResponse, error) {
```
Can be folded to 80 chars, few other instances below.

Roasbeef commented on rpcserver.go:
```diff
@@ -809,15 +814,151 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		err          error
+	)
+
+	if index < 0 {
```
Ok found the answer here. Alternatively we can use a `oneof` to say something like `PartialVerify` (which would accept an index) or `FullVerify`  (do the entire thing).

positiveblue replied:
for "partially verified"  we could do it in a reverse way

```
if len(proofs) >= index {
    // return err
}

// Proofs are store from older to newer.
proofIdx := proofFile.NumProofs() - 1 - idx

decodedProof, err = proofFile.ProofAt(proofIdx)
if err != nil {
	return nil, err
}
```

In that case you get by default the last proof, that is what you will usually want.

habibitcoin replied:
Original intention was to only decode and return a single proof, defaulting to the latest proof if index was unspecified.

Didn't think there would be much reason to return the entire proof chain

Roasbeef commented on rpcserver.go:
```diff
@@ -809,15 +814,151 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	}
+
+	var (
+		a                 = decodedProof.Asset
```
naming suggestion: `finalAsset`.

Roasbeef commented on rpcserver.go:
```diff
@@ -809,15 +814,151 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		rpcAsset          = &tarorpc.Asset{}
+		assetID           = a.Genesis.ID()
+		anchorTxId        = decodedProof.AnchorTx.TxHash().String()
+		anchorOutputIndex = strconv.FormatUint(uint64(decodedProof.InclusionProof.OutputIndex), 10)
```
Line length.

Roasbeef commented on rpcserver.go:
```diff
@@ -809,15 +814,151 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		assetID           = a.Genesis.ID()
+		anchorTxId        = decodedProof.AnchorTx.TxHash().String()
+		anchorOutputIndex = strconv.FormatUint(uint64(decodedProof.InclusionProof.OutputIndex), 10)
+		anchorOutpoint    = anchorTxId + ":" + anchorOutputIndex
```
Can instead make a `wire.OutPoint`, then call `String()` on that.

Roasbeef commented on rpcserver.go:
```diff
@@ -809,15 +814,151 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		anchorOutpoint    = anchorTxId + ":" + anchorOutputIndex
+	)
+
+	var anchorTxBuf bytes.Buffer
```
Can move this out into a serialization function similar to how things are above. Also could use some comments in the spirit of our lnd contribution guidelines: https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#code-documentation-and-commenting

habibitcoin replied:
Should be cleaned up by leveraging the already-in-place marshalAsset()

Roasbeef commented on rpcserver.go:
```diff
@@ -809,15 +814,151 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		rpcAsset.AssetGroup.RawGroupKey = a.GroupKey.RawKey.PubKey.SerializeCompressed()
+	}
+
+	resp := &tarorpc.DecodeProofResponse{
```
Can do `return &tarorpc.DecodeProofResponse{`

Roasbeef commented on cmd/tarocli/proofs.go:
```diff
@@ -72,6 +75,56 @@ func verifyProof(ctx *cli.Context) error {
+		return nil
+	}
+
+	filePath := lncfg.CleanAndExpandPath(ctx.String(proofPathName))
```
👍

Roasbeef reviewed (commented):
Nice work! This should be super useful for future UIs and also general debugging and tinkering. I think the only thing this is missing is a basic integration tests, which can cover things like:
  * the error paths (no proof, index too large, etc) 
  * proof matching up with the same proof decoded externally

positiveblue commented on rpcserver.go:
```diff
@@ -809,15 +814,151 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
 	}, nil
 }
 
+// DecodeProof attempts to decode a given proof file that claims to be anchored
```
q: is there any difference between what we call "raw poofs" and "proof files"? are they both supposed to be "raw proof in bytes"?

habibitcoin replied:
Proof file includes a genesis_point for some reason, so I guess "raw proof" is more correct than "proof file"

habibitcoin replied:
Hmm nevermind actually.. proof file is used here for consistency with the already existing VerifyProof comment

```
// VerifyProof attempts to verify a given proof file that claims to be anchored
// at the specified genesis point.
```

So I think the only difference is that a proof file also has a string field for genesis_point, but weirdly, I don't notice that string value _ever_ being populated anywhere

I also see a todo in verifier.go:
```
// TODO(roasbeef): pass in the expected genesis point here?
```

habibitcoin replied:
@Roasbeef thoughts here if genesis_point should be included?

positiveblue commented on rpcserver.go:
```diff
@@ -809,15 +814,151 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	in *tarorpc.DecodeProofRequest) (*tarorpc.DecodeProofResponse, error) {
+
+	if len(in.RawProof) == 0 {
+		return nil, fmt.Errorf("proof file must be specified")
```
nit: we can directly specify the rpc field.

habibitcoin replied:
Prefer leaving unchanged to stay consistent with VerifyProof

positiveblue commented on rpcserver.go:
```diff
@@ -809,15 +814,151 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		return nil, fmt.Errorf("unable to encode genesis: %w",
+			err)
+	}
+	rpcAsset = &tarorpc.Asset{
```
can [`marshalAsset`](https://github.com/lightninglabs/taro/blob/main/rpcserver.go#L406) help here?

if we need a subset we can create a `marshalHumanReadableAsset` helper

positiveblue commented on cmd/tarocli/proofs.go:
```diff
@@ -72,6 +75,56 @@ func verifyProof(ctx *cli.Context) error {
+
+	switch {
+	case ctx.String(proofPathName) == "":
+		_ = cli.ShowCommandHelp(ctx, "decode")
```
nice

positiveblue commented on cmd/tarocli/proofs.go:
```diff
@@ -72,6 +75,56 @@ func verifyProof(ctx *cli.Context) error {
+		return fmt.Errorf("unable to read proof file: %w", err)
+	}
+
+	resp, err := client.DecodeProof(ctxc,
```
function format

```
resp, err := client.DecodeProof(
        ctxc, &tarorpc.DecodeProofRequest{
			RawProof:   rawFile,
			ProofIndex:  ctx.Int64(proofIndexName),
        },
)
```

positiveblue reviewed (commented):
Looking good @habibitcoin :tada: 


I left some doubts and a couple of  nits

dstadulis commented:
@habibitcoin can you check positiveblue's comments? Would love to include your PR in the `taro v0.2` release.

habibitcoin commented:
@dstadulis almost done! Have addressed most comments locally. Mainly just need to add comments and tests now.

@positiveblue @Roasbeef any guidance on tests you can provide? My initial plan was to create a proofs_test.go file in itest and use assets_test as a bit of inspiration. Mint an asset, export it, verify the proof, and confirm the output matches that of the assets list CLI. Thoughts?

habibitcoin commented:
All done! Only thing is I'm struggling a bit on how/where to write tests. Similar functions, Import/Export proof do not have any integration tests so I'm a bit unsure on necessity, and also still unsure on pattern. Appreciate any guidance here!

Everything else should be addressed

guggero commented on tarorpc/taro.proto:
```diff
@@ -592,8 +598,35 @@ message ProofFile {
 
-message ProofVerifyResponse {
+message DecodedProof {
+    // The index of the decoded proof
```
nit: comments should always be complete sentences (including punctuation).

guggero commented on tarorpc/taro.proto:
```diff
@@ -592,8 +598,35 @@ message ProofFile {
+    DecodedProof decoded_proof = 2;
+}
+
+message DecodeProofRequest {
```
nit: put the request before the response?

guggero commented on rpcserver.go:
```diff
@@ -936,12 +941,118 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+
+// DecodeProof attempts to decode a given proof file that claims to be anchored
+// at the specified genesis point.
+func (r *rpcServer) DecodeProof(ctx context.Context,
```
This needs to be squashed with the previous commit, otherwise that commit won't compile on its own.

guggero commented on rpcserver.go:
```diff
@@ -936,12 +941,118 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+
+// marshalProofFile turns a proof file into an RPC DecodedProof.
+func marshalProofFile(proofFile proof.File,
+	index int64) (*tarorpc.DecodedProof, error) {
```
nit: add empty line after multi-line function definition.

guggero commented on rpcserver.go:
```diff
@@ -936,12 +941,118 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+
+	var txMerkleProofBuf bytes.Buffer
+	if err := decodedProof.TxMerkleProof.Encode(&txMerkleProofBuf); err != nil {
+		return nil, fmt.Errorf("unable to encode serialized Bitcoin merkle proof: %w",
```
nit: line length (remember to set your IDE to use 8 spaces for the tab character).

guggero reviewed (commented):
Nice work, this will definitely be useful.

> Only thing is I'm struggling a bit on how/where to write tests. Similar functions, Import/Export proof do not have any integration tests so I'm a bit unsure on necessity, and also still unsure on pattern.

I don't think we need specific test cases for this RPC. But it would be nice to at least invoke them in the existing tests, to make sure everything works end to end. So my suggestion would be to add a call to `DecodeProof` in the `verifyProofBlob()` function in `itest/assertions.go`, directly after the call to `VerifyProof`. We can do a few very basic checks on it there (no error is returned, the asset is set and so on).

habibitcoin commented:
@dstadulis ready for final review!

guggero commented on rpcserver.go:
```diff
@@ -915,10 +920,10 @@ func (r *rpcServer) DecodeAddr(_ context.Context,
-// VerifyProof attempts to verify a given proof file that claims to be anchored
-// at the specified genesis point.
+// VerifyProof attempts to verify a given proof file that claims to be
+// anchored at the specified genesis point.
```
This doesn't need to be changed, looks like the comment fit the 80 characters before.

guggero commented on rpcserver.go:
```diff
@@ -936,12 +941,128 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		return nil, fmt.Errorf("unable to encode serialized Bitcoin "+
+			"merkle proof: %w", err)
+	}
+	txMerkleProofBytes := txMerkleProofBuf.Bytes()
```
Do we need an intermediate variable here and below? Looks like we have enough space in the final struct initialization to just do:

```
	return &tarorpc.DecodedProof{
		ProofIndex:      uint64(index),
		NumberOfProofs:  uint64(proofFile.NumProofs()),
		Asset:           rpcAsset,
		TxMerkleProof:   txMerkleProofBuf.Bytes(),
		InclusionProof:  inclusionProofBuf.Bytes(),
		ExclusionProofs: exclusionProofs,
	}, nil
```

guggero commented on rpcserver.go:
```diff
@@ -936,12 +941,128 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	for _, exclusionProof := range decodedProof.ExclusionProofs {
+		var exclusionProofBuf bytes.Buffer
+		if err := exclusionProof.Encode(&exclusionProofBuf); err != nil {
+			return nil, fmt.Errorf("unable to encode exclusion proofs: %w",
```
nit: need to break before word "proofs" to fit properly (with tab with set to 8 spaces).

guggero commented on cmd/tarocli/proofs.go:
```diff
@@ -28,6 +29,9 @@ var proofCommands = []cli.Command{
 	proofPathName = "proof_file"
+
+	proofIndexName = "proof_index"
+	allProofsName  = "all_proofs"
```
This should probably match the RPC's name (`latest_proof`) since only one proof is validated.

guggero commented on cmd/tarocli/proofs.go:
```diff
@@ -72,6 +76,70 @@ func verifyProof(ctx *cli.Context) error {
+		},
+		cli.BoolFlag{
+			Name:  allProofsName,
+			Usage: "Group asset balances by group key",
```
Flag name is wrong.

habibitcoin commented:
Thanks @guggero , all points addressed! Ready again

habibitcoin commented:
For the failed tests, 1 of them is related to postgres and seems unrelated?

2 of them seem odd:
```
if a.GroupKey != nil {
		rpcAsset.AssetGroup = &tarorpc.AssetGroup{
			RawGroupKey:     a.GroupKey.RawKey.PubKey.SerializeCompressed(),
```

It is thinking that a.GroupKey.RawKey.PubKey is nil. But should that ever be possible? For a.GroupKey.RawKey.PubKey to be nil while a.GroupKey is not nil? Wondering if bug in my code, or if the test itself needs to be reimplemented properly

positiveblue commented on tarorpc/taro.proto:
```diff
@@ -592,8 +598,39 @@ message ProofFile {
+    // The raw proof in bytes to decode, which may contain multiple proofs.
+    bytes raw_proof = 1;
+
+    oneof filter {
```
we can avoid this complex field if we use the default value in our favour. Given that the default value is `0` while most of the time the endpoint will be used to decode the latest proof (which matches what the api should do even if I use it without reading the docs) what do you think about having the field `proofAtDepth` instead of `index`?

`marshalProofFile` needs to then transform this to an index with `index := proofFile.NumProofs() - 1 - proofAtDepth`, if that value is negeative it should fail. If I want to genesis proof and there are 5 proofs I need to used `proofAtDepth=4`

By default `proofAtDepth` will be 0, which means decodes the latest proof. It's a similar concept of what happens with git commits or block confirmations.

habibitcoin replied:
I like that, makes the API more self-explanatory . Great idea!

positiveblue commented on cmd/tarocli/proofs.go:
```diff
@@ -72,6 +76,70 @@ func verifyProof(ctx *cli.Context) error {
+	defer cleanUp()
+
+	switch {
+	case ctx.String(proofPathName) == "":
```
nit: you can use `!ctx.IsSet(proofPathName)`

positiveblue reviewed (commented):
Great work @habibitcoin :tada:  I think we are almost done.


I think that using `proofAtDepth` instead of `index` in the API will simplify its use + implementation. For the final pass we need to also be sure that we add godoc to all the fields and functions.

You can clean the conflicts by rebasing + regenerating the protos. The `postgres` failure is an issue that does not belong to this PR. The second issue (`a.GroupKey.RawKey.PubKey`) is something that we may need to add to the test data :+1:

habibitcoin commented on rpcserver.go:
```diff
@@ -936,12 +941,126 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		return nil, fmt.Errorf("unable to marshal proof: %w", err)
+	}
+
+	return &tarorpc.VerifyProofResponse{
```
right exactly, full proof is verified, but we only return the latest proof

not sure if we want to include an optional for previous_witnesses or a similar bool name. I was worried that it could become a massive response tho if the asset has deep history

habibitcoin commented on tarorpc/taro.proto:
```diff
@@ -592,8 +598,39 @@ message ProofFile {
+    // The index of the decoded proof.
+    uint64 proof_index = 1;
+
+    // The total number of proofs contained in the raw proof.
```
They both inherit decoded_proof, isn't that correct?

habibitcoin commented:
Just a quick update here, I'm struggling to get the test to pass. I've identified the cause is that the a.GroupKey.RawKey in specific is nil in the test case when minting an asset with emissions enabled.

But when I actually test this out with a newly minted asset, RawKey is NOT nil

```
      "asset_group": {
                "raw_group_key": "02c2f562b11b4bbb977072ac3945b5ff13551ad545a2e8f25e6777e685da08111e",
                "tweaked_group_key": "02a331696d381323589904d7a01e7a3066d934fb7c8d35024e686cfd9dd8c4ed7f",
                "asset_id_sig": "f92e77833153ccd87b365942c5e446932d363f45b112a9d336e119ed0cd4e855e38371d69e3d1e8435fa86d9c50f586ea789d11758474a52c491a9f8839636e2"
            },
            ```

I can't figure out how to get the test ensures this :(

habibitcoin commented:
I did a little more investigation and have narrowed it down to NOT being a bug with the tests, but I believe an actual bug when exporting a proof.

For some reason when you list the assets in CLI, RawKey is not nil. But when we export the proof, import it, and try to decode, RawKey is nil

lightninglabs-deploy commented:
@positiveblue: review reminder
@habibitcoin, remember to re-request review from reviewers when ready

amovfx commented:
I really like this feature and would like to use it in my testing. I'm having some trouble with this.
I mint an asset, then I go to export an asset providing the asset id and script key, and I get an error,
```
unable to find proof
```
I'm going to do some more debugging, but I think I'm missing a step

amovfx commented:
Just took this for a spin, looks like it works. I wrote a file and then read a file.I would like the meta data to be human readable.
@habibitcoin Have you used the decode to take in the raw proof data?

lightninglabs-deploy commented:
@positiveblue: review reminder
@habibitcoin, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@positiveblue: review reminder
@habibitcoin, remember to re-request review from reviewers when ready

straumat commented:
There is something that should maybe be explained in the documentation, it's about `raw_proof` parameter meaning. Sorry if I'm wrong and I did not get something.

After minting an asset, if I run:
``` 
tarocli proofs export  --asset_id 18c4eaca8905cad3ae027494b8c70a085e9452226ddea8eacfdaf6cb88431c12 \
                        --script_key 02102a7c269bf61b587c6278206d05433fa09b9fbe30d618a2cc61868622f5518f
```

I get this result:
```
{
    "raw_proof": "0000000001fd03f1002488d6b6d1f7fe8dd4134e699037b52fa7273674e3a181deacc52bdddbf62274ac00000000015000000020dbc0a8738a77d0b1f29df61820e015784d87b13f508d07af1200000000000000e9a0e62808024419f29e12075ee7192283d45787ba4e510dcc4325c4f13e2da617252c64ffff001dee9af8c502ea0200000000010188d6b6d1f7fe8dd4134e699037b52fa7273674e3a181deacc52bdddbf62274ac0000000000ffffffff02e80300000000000022512068b8ab655c6f3edfc0b241a3c560355146ed6bbf92514c09b10a0c86ac140e23910a000000000000160014c35088956d26c44af4c00dfc8bf10d9a4c79320e0247304402201829c56dba3c2ab1cd85de1fe3d7bfec8dffc2b2d4aac8c362cae25052a420290220506b92ca45a514a8b2daabb1eb8adf5283ab29b382d0331c9c884f8c89dd830f012103219e2594f427e338690c3f51409706439d8ab83fde02d18d0c28ec55c285801c0000000003fd0102083c7c4198be914f1cab7319592f0249d727fa87890ec4185b8b7881d777158d0af976bfe9e708afddbef3616fe086f90e0879295f231ac2c6d5be387ba682903bc9f67bd23cff72af821bc6feb5995b83b603aba9074ce1c74a4e13a45f108713102ce2cdd7590df19cc6c9b0f631097a2dedec3e30c1633300ea02e179eea39ef366c1a3711657d6302e516e3218b8de4a0089ca0f92621a8a49a8790dc9a1deba788b11327b109130ca4dede1a29d6ac6c03a86454199c5bc6e8a74d32df7f93d7b24f5dec13ec57f232b95d712d5947b04785cac15d00e4ea9ae6e0ffd9b3361cab39effb1b072f0918fd3ea9e9e0ae35ee61ed51e9d09548c619f06d4caa8c104e4000100014588d6b6d1f7fe8dd4134e699037b52fa7273674e3a181deacc52bdddbf62274ac000000000c6d79526f796c6c6f436f696e0e5573656420627920526f796c6c6f00000000000201000303fd03e7066901670065000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008020000092102102a7c269bf61b587c6278206d05433fa09b9fbe30d618a2cc61868622f5518f059f000400000000012102bf9d0eac1be77456ced581721d9f8837b42346173556c418794e0347e08f817102740049000100012018c4eaca8905cad3ae027494b8c70a085e9452226ddea8eacfdaf6cb88431c1202220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff012700010001220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0ea5f7971d2fe0f27d8975f656332a5f6bf0176b998a52dedaedf2ffbaef922f",
    "genesis_point": ""
}
```

BUT, if I use the previous `raw_proof` parameter value with the decode method developed by @habibitcoin, i get an error (invalid checksum).

In fact, I must export proof as a file first (with the `--proof_file` parameter) and use base 64 like this: `cat myProof | base64 -w 0`
. With this value if it works.

I think it's quite confusing for developer reading the documentation to have two parameters with same name but different meaning.

lightninglabs-deploy commented:
@positiveblue: review reminder
@habibitcoin, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@positiveblue: review reminder
@habibitcoin, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@positiveblue: review reminder
@habibitcoin, remember to re-request review from reviewers when ready

Roasbeef commented:
Hey @habibitcoin are you down to rebase this? A lot of naming stuff has changed, so that'll likely be the biggest conflict you run into locally.

Roasbeef commented:
> BUT, if I use the previous raw_proof parameter value with the decode method developed by @habibitcoin, i get an error (invalid checksum).



I think the difference is if it's just that state transition, or include the normal proof file headings.

lightninglabs-deploy commented:
@positiveblue: review reminder
@habibitcoin, remember to re-request review from reviewers when ready

habibitcoin commented:
@Roasbeef yessir! Already have it underway locally, should have it done in shortly

habibitcoin commented:
Updated and ready for another round of review!

```
Tapd % tapcli --network testnet --tlscertpath tls.cert  --macaroonpath data/testnet/admin.macaroon proofs decode --proof_file test.json
{
    "decoded_proof":  {
        "proof_at_depth":  "1",
        "number_of_proofs":  "2",
        "asset":  {
            "version":  0,
            "asset_genesis":  {
                "genesis_point":  "fde4b6b44df8e4bf8a8a48961d2a61f432146c7f742966945216208bda14dadd:1",
                "name":  "Habibitcoins",
                "meta_hash":  "49f7b90f9eefe6a5e5c6fe3e1317e1894ad721f9c9f2ac0070a631e7b3bf0302",
                "asset_id":  "3c59caaeddd621f1b537d2329aebec2087c716623395a7b72cdcf4a96702b269",
                "output_index":  0,
                "version":  0
            },
            "asset_type":  "NORMAL",
            "amount":  "20999998",
            "lock_time":  0,
            "relative_lock_time":  0,
            "script_version":  0,
            "script_key":  "02ebd16b59524d98be070002493c9724d03f68bc57c27176ac5cb1233d6c9b28ce",
            "script_key_is_local":  false,
            "asset_group":  null,
            "chain_anchor":  {
                "anchor_tx":  "0200000000010295bfdee8c817a893f6ad8af7a208e522325ffcb9a8ef6c041952e29b572894720100000000ffffffff95bfdee8c817a893f6ad8af7a208e522325ffcb9a8ef6c041952e29b5728947200000000000000000003e803000000000000225120a6a5e1a951faf32eba5592234343118a071f35dbeb1470a0cbf191fd00e730a2e8030000000000002251206e37557177f706b67fea6be30f4e6abec3fdd57358bcc67a5cd6fb30e496bc1811663900000000002251205d3bea984c8f60995b68a2533ac75ba1c1ae3f498f479c9804c9e0aa6b1e64d70140e5476d1c78d070c7cdb5ed331be125410bc9b75d8610092a9b64f146f2dffcac3e8406b930529d1f337fc5a9c22e8057f77e2398d950ce34ea6c3fcd6aec5206014037ce813bc7e3c51e260316835092d52e3a1a60f5337040470faebc86b0cbcb3e79f8cd2b81279091a28fac6206ec1eddf3892f1f5413fc8cf6a50fd157195f1a00000000",
                "anchor_txid":  "d77db8788b2613882fc102eefb3971597398631b6928ba458b2791acef03c1c3",
                "anchor_block_hash":  "00000000000092e4ac7e679a3bf8b43870bcf35cea0bcdf7e46f0531296f7e7b",
                "anchor_outpoint":  "d77db8788b2613882fc102eefb3971597398631b6928ba458b2791acef03c1c3:0",
                "internal_key":  "0256f3f99616e881163a6f6d15e6866a4800106523827c4649587d95b226e394a1",
                "merkle_root":  "385cee752b24d433680c3c346e8220535de64f99c78fc3f4c0d2447b10e2e772",
                "tapscript_sibling":  ""
            },
            "prev_witnesses":  [],
            "is_spent":  false
        },
        "tx_merkle_proof":  "070c09bbd9ca188bd8541b2de2945b90cd7f3b6ccd90223ef6b6cd11704c111d5031c054611c06bc4349e5f3f74a88df6adb8a76deadf6c045e2f2eacee03049354fb1377b7fa07a7b4f6b625a3733da0dbbb15f2560f961f94a79c814530e9515097bc160268b501230e07e8d0b79bcdeb0151825136aee74f2db70fe6657c38d1b95d11d96b440a961eb272471265dd7736ebd6e11faa7881513e80d60ba116fa5ece87e70abca83f8daa9fbe4be8fee9f2ea170a1f88e9160e415efd5c88e79aba8971a11505001629f96e2468f2b1b24e4e07ded0b8c3d486c850c30fbb1bc6c",
        "inclusion_proof":  "00040000000001210256f3f99616e881163a6f6d15e6866a4800106523827c4649587d95b226e394a10274004900010001203c59caaeddd621f1b537d2329aebec2087c716623395a7b72cdcf4a96702b26902220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff012700010001220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
        "exclusion_proofs":  [
            "000400000001012102e5a4da38c35ad44c0edcf1d007fd307584d02deb712347a3c8b356ec6264f27a029c007100010001203c59caaeddd621f1b537d2329aebec2087c716623395a7b72cdcf4a96702b269024a000140adb74259c2ab240034fcf23c67b111f21b82aefe04da57e35283612f12ef230000000000000002ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff7f012700010001220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
            "000400000002012102f59c2086590b3a04dcedd2c5a0b6565f17233abd84eab00ad382f9b6fb6508880303020101"
        ],
        "split_root_proof":  ""
    }
}
```

guggero commented on itest/assertions.go:
```diff
@@ -68,7 +68,7 @@ func assetAnchorCheck(txid, blockHash chainhash.Hash) assetCheck {
 		}
 
-		if !bytes.Equal(a.ChainAnchor.AnchorBlockHash, blockHash[:]) {
+		if a.ChainAnchor.AnchorBlockHash != blockHash.String() {
```
This commit removes the `bips` sub-directory, which I don't think is intentional.

guggero commented on rpcserver.go:
```diff
@@ -1062,12 +1062,127 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	}
+
+	decodedProof, err := r.marshalProofFile(
+		ctx,
```
nit: can write more compact, putting as many arguments on a single line up to the 80 char max.

guggero commented on rpcserver.go:
```diff
@@ -1062,12 +1062,127 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		err          error
+	)
+
+	// Default to latest proof
```
nit: missing fulls top in comment.

guggero commented on rpcserver.go:
```diff
@@ -1062,12 +1062,127 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	// Default to latest proof
+	index = int64(proofFile.NumProofs()) - 1 - index
+
+	decodedProof, err = proofFile.ProofAt(uint32(index))
```
Can just use `:=` operator here to avoid needing the `var` statement above.

guggero reviewed (approved):
Nice, LGTM, just a couple of nits.

jharveyb reviewed (commented):
Thanks for the contribution with this! left some notes re: metadata and input verification.

dstadulis commented:
@habibitcoin we're shaping the queue of PRs for `tap-v0.2.1` do you think you can update this PR to make it in?

lightninglabs-deploy commented:
@roasbeef: review reminder
@positiveblue: review reminder
@habibitcoin, remember to re-request review from reviewers when ready

habibitcoin commented on rpcserver.go:
```diff
@@ -1062,12 +1062,127 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	)
+
+	// Default to latest proof
+	index = int64(proofFile.NumProofs()) - 1 - index
```
What would be a superior UX?
```
[tapcli] unable to verify file: rpc error: code = Unknown desc = unable to marshal proof: invalid index 4294967295
```

The above is the current error message returned. I modified it so it returns the index you actually input, and I think this is most helpful to the user instead of making any assumptions about the proper index:
```
[tapcli] unable to verify file: rpc error: code = Unknown desc = unable to marshal proof: invalid index 2 is greater than latest proof index of 1
```

guggero replied:
I think it's okay the way it is, but maybe we should make the `proof_at_depth` field `uint32` to avoid it being negative? And I don't think files with more than 2 billion proofs will ever exist, so 32bit should be enough.

habibitcoin commented:
@dstadulis ready for final review!

guggero commented on rpcserver.go:
```diff
@@ -1079,12 +1079,140 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		return nil, fmt.Errorf("unable to decode proof file: %w", err)
+	}
+
+	decodedProof, err := r.marshalProofFile(ctx, proofFile, in.ProofAtDepth,
```
nit: formatting of multi-line function call. Should look like this instead:
```
	decodedProof, err := r.marshalProofFile(
		ctx, proofFile, in.ProofAtDepth, in.WithPrevWitnesses,
		in.WithMetaReveal,
	)
```

guggero commented on rpcserver.go:
```diff
@@ -1079,12 +1079,140 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	error) {
+
+	if index > int64(proofFile.NumProofs())-1 {
+		return nil, fmt.Errorf("invalid index %d is greater than latest proof "+
```
nit: line too long (should be capped at 80 characters, while counting the tab character as 8 spaces which might require you to re-configure your IDE).

guggero commented on rpcserver.go:
```diff
@@ -1079,12 +1079,140 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	index int64, withPrevWitnesses, withMetaReveal bool) (*taprpc.DecodedProof,
+	error) {
+
+	if index > int64(proofFile.NumProofs())-1 {
```
Extract `lastProofIndex := int64(proofFile.NumProofs() -1 )` as it is used 3 times here?

guggero commented on rpcserver.go:
```diff
@@ -1079,12 +1079,140 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+
+// marshalProofFile turns a proof file into an RPC DecodedProof.
+func (r *rpcServer) marshalProofFile(ctx context.Context, proofFile proof.File,
+	index int64, withPrevWitnesses, withMetaReveal bool) (*taprpc.DecodedProof,
```
Rename the `index` parameter to `depth` now as well as its meaning has changed?

habibitcoin replied:
Actually nevermind, makes sense. More consistent with CLI

guggero commented on rpcserver.go:
```diff
@@ -1079,12 +1079,140 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+
+	if withMetaReveal {
+		if len(rpcAsset.AssetGenesis.MetaHash) > 0 {
+			assetMeta, err := r.FetchAssetMeta(ctx, &taprpc.FetchAssetMetaRequest{
```
nit: line length.

guggero commented on rpcserver.go:
```diff
@@ -1079,12 +1079,140 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+					MetaHash: rpcAsset.AssetGenesis.MetaHash,
+				},
+			})
+			if assetMeta != nil && err == nil {
```
I don't think we should swallow the error here. If we specify with meta reveal but cannot fetch the meta information (because we might not have it), we should error out instead.

habibitcoin replied:
makes sense, fewer assumptions for the user

guggero commented on rpcserver.go:
```diff
@@ -1079,12 +1079,140 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	}
+
+	if withMetaReveal {
+		if len(rpcAsset.AssetGenesis.MetaHash) > 0 {
```
I think we should also error out if we request meta reveal but don't have a meta hash. That would also remove one level of indentation, e.g.:
```
if withMetaReveal {
    if len(rpcAsset.AssetGenesis.MetaHash) == 0 {
        return nil, fmt.Errorf(....)
    }

    // fetch meta info ...
}
```

guggero commented on taprpc/taprootassets.proto:
```diff
@@ -668,8 +674,58 @@ message ProofFile {
+
+    // The merkle proof for AnchorTx used to prove its
+    // inclusion within BlockHeader.
+    bytes tx_merkle_proof = 5;
```
Is the tx merkle proof really useful without the anchor TX and the block header?
I think we should add those two fields, each just the raw, serialized content.

habibitcoin replied:
IIUC, that information is already contained within "chain_anchor": {

See https://habibi.cash/asset/AdamCoin5/proof for example

guggero replied:
Not really, as we only have the block hash in there. But with the block hash someone wanting to verify the proof outside of `tapd` could fetch the whole block and do it that way, so I guess it's okay for now.

guggero commented on taprpc/taprootassets.proto:
```diff
@@ -668,8 +674,58 @@ message ProofFile {
+    // An optional TaprootProof needed if this asset is
+    // the result of a split. SplitRootProof proves inclusion of the root
+    // asset of the split.
+    bytes split_root_proof = 8;
```
There is also the `AdditionalInputs` field. But those are whole proof files (with each a list of proofs inside), which would make this too big. But maybe just `num_additional_inputs` as an information and if the user needs to know what's inside those, they can decode them themselves recursively?

guggero commented on taprpc/taprootassets.proto:
```diff
@@ -668,8 +674,58 @@ message ProofFile {
+    // the result of a split. SplitRootProof proves inclusion of the root
+    // asset of the split.
+    bytes split_root_proof = 8;
+}
```
Another field that was added recently is the `ChallengeWitness`. That can just be encoded as `repeated bytes challenge_witness = X`.

habibitcoin replied:
cool, thank you. Makes sense why this is needed for off-chain transfers. Cool!

guggero commented on cmd/tapcli/proofs.go:
```diff
@@ -81,6 +86,68 @@ func verifyProof(ctx *cli.Context) error {
+var decodeProofCommand = cli.Command{
+	Name:        "decode",
+	ShortName:   "d",
+	Description: "decode a Taproot Asset proof",
```
This should be the `Usage` field instead. See recent changes here: https://github.com/lightninglabs/taproot-assets/pull/359/commits/1607bf06d0801c3ebc72d383a7e34980376c4029

guggero commented on rpcserver.go:
```diff
@@ -1079,12 +1079,140 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+		}
+	}
+
+	rpcAsset, err = r.marshalChainAsset(ctx, &tapdb.ChainAsset{
```
nit: can just use `:=` syntax here and remove the `rpcAsset` from the `var` block above.

guggero commented on rpcserver.go:
```diff
@@ -1079,12 +1079,140 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	}
+
+	// Default to latest proof.
+	index = int64(proofFile.NumProofs()) - 1 - index
```
I think this check should be moved to the `RPC` method, since `marshalProofFile` should just contain marshaling logic and no verification IMO.

guggero reviewed (commented):
Thank you very much for the update!
Did a final pass, mostly nits, nothing major. So this is very close IMO.

habibitcoin commented:
@dstadulis ready for review again 👍

positiveblue commented on rpcserver.go:
```diff
@@ -1079,12 +1079,144 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	latestProofIndex := uint32(proofFile.NumProofs() - 1)
+
+	if in.ProofAtDepth > latestProofIndex {
+		return nil, fmt.Errorf("invalid depth %d is greater than latest proof"+
```
small nit: break the line after the space

```
		return nil, fmt.Errorf("invalid depth %d is greater than latest proof  "+
			" index of %d", in.ProofAtDepth, latestProofIndex)
```

habibitcoin replied:
Goes over the 80 char limit that way :/

guggero replied:
It still does this way (remember to configure your IDE to use 8 spaces for the tab character). So just break before the word "latest"?

positiveblue commented on rpcserver.go:
```diff
@@ -1079,12 +1079,144 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+			" index of %d", in.ProofAtDepth, latestProofIndex)
+	}
+
+	// Default to latest proof.
```
:+1:

positiveblue commented on rpcserver.go:
```diff
@@ -1079,12 +1079,144 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
+	for _, exclusionProof := range decodedProof.ExclusionProofs {
+		var exclusionProofBuf bytes.Buffer
+		if err := exclusionProof.Encode(&exclusionProofBuf); err != nil {
+			return nil, fmt.Errorf("unable to encode exclusion "+
```
would be helpful to add the index of the range in this error?

habibitcoin replied:
Like this?
```
for i, exclusionProof := range decodedProof.ExclusionProofs {
		var exclusionProofBuf bytes.Buffer
		if err := exclusionProof.Encode(&exclusionProofBuf); err != nil {
			return nil, fmt.Errorf("unable to encode exclusion "+
				"proof of index %d: %w", i, err)
		}
		exclusionProofBytes := exclusionProofBuf.Bytes()

		exclusionProofs = append(exclusionProofs, exclusionProofBytes)
	}
	```

positiveblue reviewed (approved):
Thank you for pushing this to the end @habibitcoin :tada:  

LGTM. I see that you added some of the fields that @guggero proposed to the grpc messages but they are not used anywhere. Do you plan to populate the `num_additional_inputs`/`challenge_witness` fields in your responses too??

habibitcoin commented:
@positiveblue thanks for the review! Not sure I understand your comment re; the addtl fields, they should be returned properly I believe, attached a sample output:
```
{"decoded_proof":{"proof_at_depth":1,"number_of_proofs":2,"asset":{"version":0,"asset_genesis":{"genesis_point":"8f9dea9fbfd22c9c4084fc922650652bab0840eac6d3e139b45d1f1ddb35757d:2","name":"PythonBux","meta_hash":"799b3ea0b60e9c4c66dfeb552825db6d3b7bab5460bc6f07734aae72f70671c4","asset_id":"54df2137150ddddd9357623069418789242ae3bf5a8b7b6be8a67577a14be14e","output_index":0,"version":0},"asset_type":"NORMAL","amount":"1","lock_time":0,"relative_lock_time":0,"script_version":0,"script_key":"0206667113ba0b00e8f284abf6941b37e2b14326c4289cadb010db0f0cf89b21c0","script_key_is_local":false,"asset_group":null,"chain_anchor":{"anchor_tx":"02000000000102c853219c7743adc1bb311019edb4e8e7eb75f83b29e99dc698bb680f03bf0ce30100000000ffffffffc853219c7743adc1bb311019edb4e8e7eb75f83b29e99dc698bb680f03bf0ce300000000000000000003e803000000000000225120f5ea05c05bfe1b3a72f17ddaff83e608cacdbef4e3e0a88441e1c503961f541ae803000000000000225120323e8392eb255ca02b64202528678dddef7541d6b59e3c2cee455c377e731f11a81d13000000000022512044dd1493ccfa2e3cc6e526ac82cc1c201a18ece07345727b08b789314a6409a90140457fcedd8a3de9ae574f1de6f310dbe02a4324011279e77cc6bebe928c76f0095a520c558efa6994f0f5721fad6b4ba50fc2d42e6f9301cfd8303ee3422d21ef0140dcdb3e79f785f053493e2e0536d9622bc2f98e8c1007f944ff228d10c2c56fe1d3c402e074298d57eb96262642c76c8152e031b2b382d1eef613e1442f8f71bd00000000","anchor_txid":"8859fe78c2fc0d3d98e0a6238cb10bf87798ddd5ed994ba494680d7bb9d99c91","anchor_block_hash":"000000000000000b8016b76f70f2b562b23cb6e8247b87ff021fc746be7683a7","anchor_outpoint":"8859fe78c2fc0d3d98e0a6238cb10bf87798ddd5ed994ba494680d7bb9d99c91:1","internal_key":"02ecdae74d996a281b61c911c51b9533383b4358f10f02943ec102b1ace9dff292","merkle_root":"","tapscript_sibling":""},"prev_witnesses":[{"prev_id":{"anchor_point":"0000000000000000000000000000000000000000000000000000000000000000:0","asset_id":"0000000000000000000000000000000000000000000000000000000000000000","script_key":"000000000000000000000000000000000000000000000000000000000000000000","amount":"0"},"tx_witness":[],"split_commitment":{"root_asset":{"version":0,"asset_genesis":{"genesis_point":"8f9dea9fbfd22c9c4084fc922650652bab0840eac6d3e139b45d1f1ddb35757d:2","name":"PythonBux","meta_hash":"799b3ea0b60e9c4c66dfeb552825db6d3b7bab5460bc6f07734aae72f70671c4","asset_id":"54df2137150ddddd9357623069418789242ae3bf5a8b7b6be8a67577a14be14e","output_index":0,"version":0},"asset_type":"NORMAL","amount":"99","lock_time":0,"relative_lock_time":0,"script_version":0,"script_key":"023dba4c1cb1b66a8c27e777b9eea380b5398f1287c2d3fdcda933822eb4cb39e2","script_key_is_local":false,"asset_group":null,"chain_anchor":null,"prev_witnesses":[{"prev_id":{"anchor_point":"e30cbf030f68bb98c69de9293bf875ebe7e8b4ed191031bbc1ad43779c2153c8:0","asset_id":"54df2137150ddddd9357623069418789242ae3bf5a8b7b6be8a67577a14be14e","script_key":"0217171c348984aa542bee4e50b741d92ce6e4466f8af1fcd6467b9509c1312e26","amount":"0"},"tx_witness":["4f62b268edbd370103d6c34239bf74762f0af714c45d04c8ad35ea488ec4a4c0af2fa9b3dfa1c5cc78f77dcedf82a29dc669bc1187d8b227b21820ff2344568f"],"split_commitment":null}],"is_spent":false}}}],"is_spent":false},"meta_reveal":{"data":"","type":"META_TYPE_OPAQUE","meta_hash":"01e246b58d8e782fc96881c090d833eefa37e804cb308aeae0f7471c9ef1ea1a"},"tx_merkle_proof":"06fa2f4f9f7a5b558f21876f015ce806e878cacb3882d1d263841b628cee41909f6e74740e2bc67dd8d1a902cd53c812e2015cbf729233bbbfde8e636ed194c0394f3ba9c2621c1cb174343c9119c76dfb03d45fbad8500eb7062a3d874e85e6d7b60c5c419d44febe41ed53193630ebdc6ccf12295a913f488f915226ae746a56e0f2de349e34148a2e123ab02f6e640de4104ea4e6edb14ff3d8ed34b3a2f0eb7eac4f77ee31eadd4dafd5664777945bb6e8cdf26469b5043c8bc2511d85e2b22e","inclusion_proof":"000400000001012102ecdae74d996a281b61c911c51b9533383b4358f10f02943ec102b1ace9dff29202740049000100012054df2137150ddddd9357623069418789242ae3bf5a8b7b6be8a67577a14be14e02220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff012700010001220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff","exclusion_proofs":["00040000000001210254dce613492424e99c795c7e99d9523b97af8d3bcbfbdc3044e0ab0c9c4cbcf5029c0071000100012054df2137150ddddd9357623069418789242ae3bf5a8b7b6be8a67577a14be14e024a000152928ba8f5b20ae8f5ebad58f80c478805ce83b28aa825c2870ce4d930d96d060000000000000063ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffbf012700010001220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff","000400000002012102fb0cc7cee22cb0620e47674dfb0972977b9dc1d6713a6526863fccf6c690d2d00303020101"],"split_root_proof":"00040000000001210254dce613492424e99c795c7e99d9523b97af8d3bcbfbdc3044e0ab0c9c4cbcf502740049000100012054df2137150ddddd9357623069418789242ae3bf5a8b7b6be8a67577a14be14e02220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff012700010001220000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff","num_additional_inputs":0,"challenge_witness":[]}}
```

guggero reviewed (approved):
Very nice! Just two more small nits then we're good to go.
